### PR TITLE
BSN Syntax Improvements

### PR DESCRIPTION
**Pull request by cart**
# Objective

BSN currently treats passed in variables and inline functions as "scenes", and requires `template_value()` to pass in a component `Template` value:

```rust
fn widget(node: Node, contents: impl Scene) -> impl Scene {
    bsn! {
        Widget
        template_value(node)
        contents
    }
}
```

This is a bit unexpected, as `node` is a "peer" of `Widget`, from a placement perspective. `template_value` comes up _a lot_,  and results in significant confusion and noise.

Other "scene inclusions", such as `@SceneComponent {}` and `:"scene.bsn"` have explicit visual indicators that they are scenes.

This all results in making `bsn!` harder to write and read than it should be. We can do better.

Additionally, this pattern shows up a lot:

```rust
bsn! {
    template_value(Transform::from_xyz(-2.5, 4.5, 9.0).looking_at(Vec3::ZERO, Vec3::Y))
}
```

As does this:

```rust
bsn! {
    template_value(foo.bar())
}
```

The `template_value` wrapper is required because `bsn!` doesn't currently support expressions like `foo.bar()`. `template_value` fixes this by breaking out of `bsn!`'s syntax / reading the Rust expression directly. We should support this pattern directly and it should behave as expected!

Additionally, enums (especially enums outside of developers control) are annoying and weird to work with in BSN, as they require `VariantDefaults` to support "per variant patching". This is a nice feature, but it hasn't really come up in practice, and the cost it incurs is way too high from a usability perspective.

Enums that don't implement VariantDefaults (or implicitly via FromTemplate) require using template_value:

```rust
bsn! {
  template_value(Team::Blue)
}
```

Additionally, things like `NumberInputValue::F32` required `template_value`, as it is ambiguous whether it is an enum variant an or "associated const":

```rust
bsn! {
  template_value(NumberInputValue::F32(1.0))
}
```

So many annoyances!

## Solution

- Variables, functions, and expressions in "BSN entry" position are now interpreted as templates.
- All scene inclusions now use `@` syntax, which is now considered to be the "uncached scene" syntax.
- BSN entries now support `foo.bar()`-style expressions. These are always interpreted as "template values"
- I have removed `{}` from "BSN entry position", as `Widget {}` and `Widget { foo }` is ambiguous. This also provides clarity in the "scene list" case, as `{}` is now only used in list item position.
- I've implemented Template for the `Propagate` component so it can be used directly
- Enums now require specifying every field value. This means any enum that implement Default and Clone can now be used with BSN, just like other types. Structs defined inside enums therefore _also_ require specifying every field, as they are being written on top. However these still have "implicit default".
- Thanks to the new enum behavior, the `Type::F32` ambiguity (is it an enum or an associated const) is no longer a problem. These can both parse to the same AST and resolve to the same expression.
- "struct update" syntax is now supported
- BSN "codegen" has been refactored for improved clarity and terseness
- Code and examples have been ported
- Added support for `~{TEMPLATE_EXPRESSION}` syntax, which supports arbitrary rust-code template expressions.
- `template_value` has been deprecated. In 99.99% of cases, BSN code can remove this wrapper. On the off chance you need an arbitrary template expression that isn't BSN compatible (ex: multiple statements), you can now use `~{TEMPLATE_EXPRESSION}` syntax. 

This is how it looks now:

```rust
fn widget(node: Node, contents: impl Scene) -> impl Scene {
    bsn! {
        Widget             // component template
        node               // component template variable
        node.clone()       // function that returns a component template
        @contents          // uncached Scene variable
        @{contents}        // uncached Scene expression
        @scene_function()  // uncached Scene function
        @SceneComponent    // uncached Scene Component 
        // An expression that returns a component template
        Transform::from_xyz(-2.5, 4.5, 9.0).looking_at(Vec3::ZERO, Vec3::Y)
    }
}
```

Ultimately I think we should consider deprecating `template_value`, but we need to keep it around for a bit longer, as it is currently used in one "corner case" in an example.

### CODE REVIEW

**Comment by gagnus** [comment]
This looks excellent

File: crates/bevy_feathers/src/assets/shaders/color_plane.wesl
Comment (tmstorey) [inline crates/bevy_feathers/src/assets/shaders/color_plane.wesl:1]: This seems to have gone missing from #24678 and wound up here.

**Review by Diddykonga — APPROVED** [review]

**Review by HugoPeters1024 — COMMENTED** [review]
Thank you for this, this is massive improvement already! 

Although one gripe that remains is the whitespace vs comma separators. Consider this example:

```rust
bsn! {
    Children [
        A
        B,
        C,
    ]
}
```

It's incredibly hard to miss the comma (or lack thereof) while it impacts semantics a lot. (For those who don't realize, the above spawns 2 child entities with [A,B] and [C] components.)

This reminds me of a similar footgun that exists in python that has been the cause of at least 2 production incidents in my life:

```python
options: list[str] = [
    "A"
    "B",
    "C",
]
```

`options` will evaluate to `['AB', 'C']`

Now I'm not saying this PR has this to fix this issue to be good enough to merge, but I do fear that a proper solution requires a more fundamental syntactic overhaul. Since this PR already does that, it might be worth it to see if it can be done as part of this PR.

(This problem is closely related though somewhat distinct from an earlier issue that I brought up regarding the `children!` macro: https://github.com/bevyengine/bevy/issues/20429)

File: examples/3d/order_independent_transparency.rs [hunk 1]
```diff
@@ -263,14 +263,14 @@ fn spawn_ui(commands: &mut Commands, app_state: &AppState) {
 
     // Buttons
     commands.spawn_scene(bsn! {
-        main_ui_node_scene()
+        @main_ui_node_scene()
         // Prevent the event from bubble up so that view drag does not initiate when interacting with the UI
-        on (|mut event: On<PointerDrag>| {
+        @on (|mut event: On<PointerDrag>| {
```
Comment (qoh) [inline examples/3d/order_independent_transparency.rs]:
```suggestion
        on (|mut event: On<PointerDrag>| {
```

**Comment by alice-i-cecile** [comment]
I would like to improve that too; I think it's dangerously hard to delineate. That said, I want to get this PR merged as is, so I've spun out #25616 as a stable home for that discussion. Swapping out the entity seperator should be fairly straightforward, technically, but come with a ton of documentation changes so I don't think bundling it will make the change easier to review / do.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I very much appreciate these improvements, especially the reduction of `template_value!`. Some feedback before we merge though:

1. The table for docs in the bsn! macro was missed, and needs to be updated.
2. With the help of The Machine (Opus 5), I've found a number of broken edge cases that should be fixed before we merge. PR [here](https://github.com/cart/bevy/pull/54) with the failing tests:
3. We can actually get rid of the remaining `template_value!` call :D PR [here](https://github.com/cart/bevy/pull/53) to just fix it. With that done, IMO we should just deprecate the template_value! macro completely in this cycle.

**Comment by cart** [comment]
> The table for docs in the bsn! macro was missed, and needs to be updated.

I did update most of the table already, but I missed the root and a couple cases. I've updated these!

> With the help of The Machine (Opus 5), I've found a number of broken edge cases that should be fixed before we merge. PR https://github.com/cart/bevy/pull/54 with the failing tests:

One of the tests didn't make sense, but I've fixed the other ones. Thanks!

> We can actually get rid of the remaining template_value! call :D PR https://github.com/cart/bevy/pull/53 to just fix it. With that done, IMO we should just deprecate the template_value! macro completely in this cycle.

I wasn't comfortable deprecating this as it serves as an escape hatch for arbitrary template expressions (bsn only supports "chained expressions"). Fortunately, template syntax (`~`)  already exists, so I've added support for template expressions `~{SOME_EXPR}`. That fills the same role as `template_value`, so I've deprecated `template_value`.

This should be ready to merge now!

**Review by alice-i-cecile — APPROVED** [review]
